Migrate App Router to Page Router

bradjasper opened the pull request:
This PR migrates the codebase from Next.js App Router to Pages Router to enable WebCat support.

### Why

WebCat requires strict Content Security Policy (CSP) enforcement to verify that the code running in the browser matches the open source repository. The App Router injects dynamic `<script>` tags into statically compiled pages, which violates CSP and breaks WebCat's verification model.

The Pages Router does not have this behavior—it outputs static, hashable script references that are compatible with strict CSP.

### What Changed

- **Routing structure**: Moved from `src/app/` to `src/pages/` with `_app.tsx` and `_document.tsx`
- **Layouts**: App Router layouts replaced with a single `_app.tsx` wrapper (we only had one layout anyway)
- **Error handling**: Added custom `_error.tsx` and `404.tsx` pages. Errors are now logged rather than thrown, which keeps development working without backend API access (e.g., Clerk key mismatches during local dev)
- **Clerk middleware runtime**: Changed from Edge to Node.js runtime. We're not running on edge infrastructure, and this fixes compiler warnings during production builds
- **SSE streaming**: Updated dev simulator to properly flush streaming responses. The Pages Router buffers differently than App Router, so explicit flush calls are now required
- **Removed `'use client'` directives**: These are no-ops in Pages Router

### Testing Notes

- Basic functionality tested locally
- **Needs production testing** with live Clerk API keys, R2 cloud sync, and live models before full rollout

sachaservan commented on src/pages/index.tsx:
Should /app directory be deleted? It seems like right now the build is missing html files. It looks like maybe next.js detects the /app dir and tries build a hybrid app/pages?

lsd-cat replied:
To backup this, my attempt at building statically is adding the "export" config option to the Next config. Before it was throwing, because of the dynamic routes. Now it completes successfully, but it seems like the output doesn't contain everything necessary to run it:

```
# find .
.
./static
./static/css
./static/css/209246b8fb178958.css
./static/css/3864b451a61e4546.css
./static/chunks
./static/chunks/524.8010810f85ffdf60.js
./static/chunks/c2eea924.39ff5e680ea6d65f.js
./static/chunks/58474bfb.960ab4d204ab8596.js
./static/chunks/main-8abe64ce6c3336a3.js
./static/chunks/2bd9703c.0c4c993a670df5de.js
./static/chunks/0bd2f4cc.32df12a145ffb8a3.js
./static/chunks/714-04ecf911909e2b97.js
./static/chunks/230-c206ae94ef68779e.js
./static/chunks/482.46c742b792a52cb7.js
./static/chunks/143.84cc491a6adfb87d.js
./static/chunks/webpack-d55d43433ee1af5a.js
./static/chunks/c859536c-e744ee0a046991d7.js
./static/chunks/317.b2a6aada5f6e3fdb.js
./static/chunks/839.183bb831ec3a2957.js
./static/chunks/718.e46c081aa977b272.js
./static/chunks/235.6362f5de54eec9c2.js
./static/chunks/216.7a8fbafe1a864ea2.js
./static/chunks/280.3cf0b3edf71ba3cd.js
./static/chunks/0465ed09-6e8dfea69f4e0d9b.js
./static/chunks/284.4aafb7c8552dd3f0.js
./static/chunks/6f507d93-109439022adfa87b.js
./static/chunks/899.3cf0b3edf71ba3cd.js
./static/chunks/2b38e19b-005cc927a6819311.js
./static/chunks/671.af4262e945c41a87.js
./static/chunks/b258f366.77bfc66f7fd88cb9.js
./static/chunks/489b9f9a.39077f072585002a.js
./static/chunks/618.cfee59f4244a3066.js
./static/chunks/291.bfb84d15b527b598.js
./static/chunks/606.582131d0a60a963e.js
./static/chunks/a7e16ef1.897916e04caea015.js
./static/chunks/126.2794e718ba458407.js
./static/chunks/45.6f06dd19e5983f93.js
./static/chunks/11.e8ca0733092a21fb.js
./static/chunks/602.b6eaf3a9d1e56d5f.js
./static/chunks/pages
./static/chunks/pages/index-7562bb442907924d.js
./static/chunks/pages/_error-8f2a44282e22edd9.js
./static/chunks/pages/404-94b1f90d77f3931a.js
./static/chunks/pages/_app-19a7ac1316be3385.js
./static/chunks/802.e5a365619bd8436e.js
./static/chunks/polyfills-42372ed130431b0a.js
./static/chunks/framework-a6e0b7e30f98059a.js
./static/dMNDfNqz1DAwufaBTLLmm
./static/dMNDfNqz1DAwufaBTLLmm/_ssgManifest.js
./static/dMNDfNqz1DAwufaBTLLmm/_buildManifest.js
./static/media
./static/media/KaTeX_Fraktur-Regular.ed38e79f.ttf
./static/media/KaTeX_Math-Italic.5295ba48.woff
./static/media/KaTeX_Math-Italic.f28c23ac.woff2
./static/media/KaTeX_Main-Regular.0462f03b.woff2
./static/media/KaTeX_Main-Bold.d181c465.woff
./static/media/e9b828b317d9581d-s.p.woff2
./static/media/KaTeX_SansSerif-Regular.74048478.woff
./static/media/KaTeX_Size4-Regular.10d95fd3.woff2
./static/media/KaTeX_Caligraphic-Regular.55fac258.woff2
./static/media/ba9851c3c22cd980-s.woff2
./static/media/KaTeX_AMS-Regular.4aafdb68.ttf
./static/media/KaTeX_Main-Italic.8916142b.woff2
./static/media/KaTeX_SansSerif-Bold.94e1e8dc.ttf
./static/media/KaTeX_SansSerif-Regular.ba21ed5f.woff2
./static/media/KaTeX_SansSerif-Italic.3b1e59b3.woff2
./static/media/KaTeX_Fraktur-Bold.9f256b85.woff
./static/media/KaTeX_Main-Regular.b7f8fe9b.ttf
./static/media/KaTeX_Fraktur-Regular.d3c882a6.woff2
./static/media/19cfc7226ec3afaa-s.woff2
./static/media/21350d82a1f187e9-s.woff2
./static/media/046d8042041adc5e-s.p.woff2
./static/media/KaTeX_Script-Regular.07505710.woff
./static/media/KaTeX_Caligraphic-Regular.dad44a7f.woff
./static/media/bd6b397eeece3ba9-s.p.woff2
./static/media/KaTeX_Size2-Regular.d6b476ec.ttf
./static/media/KaTeX_Fraktur-Bold.d42a5579.woff2
./static/media/KaTeX_SansSerif-Italic.7c9bc82b.woff
./static/media/KaTeX_AMS-Regular.a79f1c31.woff2
./static/media/7c004a8f86cee87f-s.p.woff2
./static/media/KaTeX_Caligraphic-Regular.07ef19e7.ttf
./static/media/KaTeX_Fraktur-Regular.7c187121.woff
./static/media/KaTeX_Math-Italic.939bc644.ttf
./static/media/KaTeX_Main-BoldItalic.e3f82f9d.woff
./static/media/KaTeX_SansSerif-Italic.b4c20c84.ttf
./static/media/b9c80ac26089c2d7-s.p.woff2
./static/media/KaTeX_Script-Regular.03e9641d.woff2
./static/media/KaTeX_Math-BoldItalic.a879cf83.ttf
./static/media/KaTeX_Script-Regular.fe9cbbe1.ttf
./static/media/KaTeX_SansSerif-Bold.bf59d231.woff
./static/media/KaTeX_SansSerif-Regular.d4d7ba48.ttf
./static/media/df0a9ae256c0569c-s.woff2
./static/media/5d0b646a3634f266-s.p.woff2
./static/media/KaTeX_Size4-Regular.fbccdabe.ttf
./static/media/KaTeX_Typewriter-Regular.d97aaf4a.ttf
./static/media/KaTeX_Size1-Regular.e1e279cb.woff
./static/media/KaTeX_Typewriter-Regular.a8709e36.woff2
./static/media/KaTeX_Main-Bold.b74a1a8b.ttf
./static/media/KaTeX_Size4-Regular.7a996c9d.woff
./static/media/KaTeX_Main-BoldItalic.70d8b0a5.ttf
./static/media/acc6cd0d95ce17b0-s.p.woff2
./static/media/KaTeX_Caligraphic-Bold.ec17d132.woff2
./static/media/KaTeX_AMS-Regular.1608a09b.woff
./static/media/c98ff9038ce9979c-s.p.woff2
./static/media/KaTeX_Caligraphic-Bold.cce5b8ec.ttf
./static/media/KaTeX_Main-Regular.7f51fe03.woff
./static/media/KaTeX_Main-Italic.47373d1e.ttf
./static/media/KaTeX_Size2-Regular.5916a24f.woff2
./static/media/4df13d456324dcb4-s.p.woff2
./static/media/KaTeX_Math-BoldItalic.f1035d8d.woff
./static/media/KaTeX_Fraktur-Bold.b18f59e1.ttf
./static/media/KaTeX_Size3-Regular.a144ef58.ttf
./static/media/8e9860b6e62d6359-s.woff2
./static/media/KaTeX_Caligraphic-Bold.b6770918.woff
./static/media/KaTeX_Size1-Regular.eae34984.woff2
./static/media/KaTeX_Main-Italic.9024d815.woff
./static/media/c5fe6dc8356a8c31-s.woff2
./static/media/KaTeX_Size3-Regular.9acaf01c.woff
./static/media/KaTeX_Size1-Regular.fabc004a.ttf
./static/media/KaTeX_SansSerif-Bold.8c5b5494.woff2
./static/media/KaTeX_Size2-Regular.57727022.woff
./static/media/KaTeX_Main-BoldItalic.6f2bb1df.woff2
./static/media/KaTeX_Typewriter-Regular.6258592b.woff
./static/media/KaTeX_Math-BoldItalic.572d331f.woff2
./static/media/e4af272ccee01ff0-s.p.woff2
./static/media/KaTeX_Size3-Regular.b4230e7e.woff2
./static/media/KaTeX_Main-Bold.c3fb5ac2.woff2
./dMNDfNqz1DAwufaBTLLmm

```

bradjasper replied:
> Should /app directory be deleted? It seems like right now the build is missing html files. It looks like maybe next.js detects the /app dir and tries build a hybrid app/pages?

`/app` should be removed from the repo in this branch — if you still see it perhaps a `git clean` or clear Next.js build cache.

> To backup this, my attempt at building statically is adding the "export" config option to the Next config. Before it was throwing, because of the dynamic routes. Now it completes successfully, but it seems like the output doesn't contain everything necessary to run it

The issue here is the backend server (middleware, API routes and headers) is kicking Next.js into dynamic Node.js mode. The build system sees these and doesn't want to spit out static HTML pages.

@sachaservan a good solution is to split the frontend and the backend so Next.js can cleanly build a 100% static frontend, while also serving a dynamic backend. This should be straightforward to do as a mono-repo, please let me know if you'd like me to proceed here!

sachaservan commented on src/utils/error-handling.ts:
```diff
@@ -21,7 +21,19 @@ export function logError(
+      localStorage.getItem('enableDebugLogs') === 'true' &&
+      error instanceof Error
+    ) {
+      console.log('Stack trace:', error.stack)
```
I know we had console.error before but might be better to take this opportunity to use the logError utility?

bradjasper commented on src/components/chat/WelcomeScreen.tsx:
```diff
@@ -196,14 +199,19 @@ export const WelcomeScreen = memo(function WelcomeScreen({
-                                      : model.image === 'moonshot.png'
-                                        ? `/model-icons/moonshot-${isDarkMode ? 'dark' : 'light'}.png`
-                                        : `/model-icons/${model.image}`
+                                    failedImages[model.modelName]
```
@sachaservan I've got these rebased on main, only changes from me are in this PR now. Will ensure I do this going forward.

bradjasper commented:
Just pushed some changes to enable fully static builds:

* Switched Next `output` to `export`
* Clerk middleware and endpoint were removed in favor of client-side access. This was possible because we use public attributes on Clerk for determining paid account status, which is available locally.
* Removed health endpoint
* Moved dev simulator to separate backend script, still tests network/SSE, but isn't included in production build
* Moved dynamic headers() to `vercel.json` file — if another host is used (like self host with nginx), headers need to be added there. It's possible to move CSP rules to HTML `<meta>` tag, but messy in dev where we don't use them, and we need the other headers anyway.

You may need to clear `.next/` cache to get builds working.

Also more testing is needed before fully deploying, as this is a pretty significant difference from how production is currently deployed.

sachaservan commented on src/hooks/use-subscription-status.ts:
```diff
@@ -1,126 +1,97 @@
+      publicMetadata['chat_subscription_expires_at'],
+    )
+
+    const now = new Date()
```
Fixing this will add more complexity than is worth it. For subscription expiration, having slight inaccuracy is fine.

sachaservan commented on next.config.mjs:
```diff
@@ -1,5 +1,12 @@
 /** @type {import('next').NextConfig} */
 const nextConfig = {
+  output: 'export',
```
The security headers were moved to vercel.json as part of the static export migration. All headers are configured there. For self-hosting with nginx or other platforms, headers would need to be configured at that layer, but that's expected for any static site deployment.

sachaservan commented:
@lsd-cat would you mind looking over the latest version on this PR, which should now be fully static, as see if it fixes all the issues related to server-side rendering from https://github.com/tinfoilsh/tinfoil-chat/issues/166?
